GitHub migrations checks

joxl opened the pull request:
## Technical Summary
Add migrations lockfile and container test for verification.

Jira: [SAAS-12801](https://dimagi-dev.atlassian.net/browse/SAAS-12801)

## Safety Assurance

### Safety story

Invisible, affects code contributors only.  Adds container check (under `python-sharded-and-javascript` check) for missed or invisible migrations.

### Automated test coverage

Appends existing `scripts/test-django-migrations.sh` to fail merge eligibility if migrations lockfile is not up-to-date.

### QA Plan

No QA.

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
- [x] The migrations in this code can be safely applied first independently of the code

There are no real migrations in this PR, but it _should_ get labeled as such because the `migrations.lock` file has changed.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

joxl committed:
Add `migrations.lock` as of c989ac7faccb600a8abb7acf91adf130239f9736

joxl committed:
Add Makefile command for updating migrations lockfile

joxl committed:
Include `migrations.lock` in pattern for GitHub migrations label rule

joxl committed:
Add GitHub migrations workflow

snopoke commented on .github/workflows/migrations.yml:
```diff
@@ -0,0 +1,21 @@
+  pull_request:
+    branches:
+    - master
+    - hotfix-deploy
```
do we ever use this branch? I would have thought it would always be a dynamic name like `hotfix-2021-12-01`

snopoke replied:
(Interesting, I see we have it documented like this on the wiki: https://confluence.dimagi.com/display/GTD/Deploy+Process)

joxl replied:
I think there is a new workflow in place to start using it more. I'll admit I just copied this from `.github/workflows/tests.yml`, but I did notice it and assumed it's valid.

snopoke commented on Makefile:
```diff
@@ -8,3 +8,12 @@ upgrade-requirements:
+
+migrations: TMPLOCK:=migrations.new.lock
+migrations:
+	./manage.py showmigrations --list > $(TMPLOCK) || (rc=$$?; rm -vf $(TMPLOCK); exit $$rc)
+	@if diff --color /dev/null /dev/null >/dev/null 2>&1; then \
+		diffcolor=--color; \
+	fi; \
+	diff $$diffcolor -su migrations.lock $(TMPLOCK) || true
```
🐚 🥷

snopoke reviewed (commented):
nice 👮🏻

joxl commented:
Reviews welcome. The "migration check" technique is complete unless reviews suggest changes.  I still need to add a working migration check in the workflow (requires getting the check to run in the docker web container).

snopoke commented:
Build fails `no module named gevent`

snopoke reviewed (changes requested):
Thinking about this more I don't think this can work here since there's not DB available. You'll either need to setup a DB or else run this inside our docker setup.

Once you have a DB you'll first need to run `./manage.py migrate` unless you strip of the `[x]` from the output.

joxl commented on .github/workflows/migrations.yml:
```diff
@@ -0,0 +1,21 @@
+        submodules: recursive
+    - name: Verify up-to-date migrations lockfile
+      run: |
+        make migrations
```
Yeah, this won't work I don't think we actually setup any env in the "workflow image". I need to add a "runner" inside the django docker container, spin up that container and kick it off. I think it will be something like:

```yaml
    - name: Test migrations
      env:
        TEST: migrations
        JS_SETUP: no
      run: scripts/docker test
```

snopoke replied:
I do really like having this as a separate workflow though rather than part of the 'test' workflow

joxl commented:
> I don't think this can work here since there's not DB available.

Correct, this kind of thing has historically been added as a function of the `scripts/docker` utility, which knows how to spin up and prepare the entire stack in containers.

millerdev commented:
I like the idea. Curious about where the motivation for doing this came from?

Unfortunately spinning up docker and services is a non-trivial process that takes a lot of time (at least 3 minutes, I think). It would probably be easiest to move this into one of the test runners. Probably python-sharded-and-javascript where we have a bunch of other similar checks like whether you need to run `make requirements`. What were the reasons to make this a separate Action?

joxl commented:
> Curious about where the motivation for doing this came from?

Upgrading Django plugins which perform migrations is an easy way for a developer to unknowingly miss that their PR contains migrations. This new PR check will prevent a developer from missing that their PR contains migrations.

> What were the reasons to make this a separate Action?

It doesn't fit into any existing actions.  The "test" functions are for python and javascript. There is also a `static_analysis` that runs under `python-sharded-and-javascript`, but that's not a "check and fail" kind of thing, it just reports metrics.

joxl commented:
> Unfortunately spinning up docker and services is a non-trivial process that takes a lot of time (at least 3 minutes, I think).

Sounds like there could be value in adding a post-merge action that builds a "latest" container image which PR jobs could use.

millerdev commented:
> could be value in adding a post-merge action that builds a "latest" container image which PR jobs could use.

It's not building images that is taking time. Currently our "web" image is rebuilt by Docker Hub each time a PR is merged into master (it's automated and doesn't block Actions or anything). Actions just grabs the most recent build available on each run. The slow part is downloading 7 or so large-ish Docker images (multiple GB in total).

Github Actions actually is somehow seemingly more efficient at this than Travis was. I concluded this using a very small sample set, so could be wrong or the story may be more nuanced than that. On Travis Docker initialization took 5-7 minutes. I'm not sure how Github is doing it, maybe they have some kind of cache layer in the data center(s) where they run actions?

The only ways I've thought of getting around this is to run less services or try to build small(er) custom Docker images for all of our services, or at least the biggest ones. Building smaller images does not appear to be a trivial task, and would require us to maintain more things.

joxl commented:
> The slow part is downloading 7 or so large-ish Docker images (multiple GB in total).

Ah, what I suggested is already being done. I thought the slowdown was the pip/yarn installs, but it's actually fetching the images. Ok, that is a strong vote against a separate action.

joxl committed:
Remove GitHub action and replace by extending existing script.

Turns out `scripts/test-django-migrations.sh` already existed and is
part of our current container test suite.

joxl commented:
I'm expecting the new `python-sharded-and-javascript` check to fail because I've intentionally left the `migrations.lock` file stale. I will push an updated one once the check fails.

joxl commented:
> Probably python-sharded-and-javascript where we have a bunch of other similar checks

Indeed, there are other checks, including one that already checks for "missed migrations" (I'm not 100% sure what that means in the scope of a container test run where migrations are explicitly run to update the DB, but anyway).  The new `make migrations` and subsequent lockfile check are now part of that script.

_Sidenote: why is it called `python-sharded-and-javascript`? I advocate that we rename it to something like `catchall` or `everything-else`._

joxl committed:
Update migrations with `make migrations`

joxl commented:
Force-push was to remove a `WIP` commit at the previous HEAD. No changes were made to previously-reviewed code.

snopoke commented on Makefile:
```diff
@@ -8,3 +8,12 @@ upgrade-requirements:
 	cd docs && $(MAKE) html; cd -
+
+migrations: TMPLOCK:=migrations.new.lock
+migrations:
```
based on the name I would expect `make migrations` to do something similar to `./manage.py makemigrations`. How about renaming this to `migrations_lock` or `migrations_list`?

millerdev replied:
+1 for `make migrations.lock` (if dots are allowed) if not, `make migrations_lock`

snopoke commented on scripts/test-django-migrations.sh:
```diff
@@ -1,11 +1,52 @@
+        return 1
+    fi
+    # clean
+    log_color '0;32' "frozen migrations ok"
```
it would be great to either alias the color codes or the function:
```
log_color "$success" "success message"
# OR
log_success "success message"
```

snopoke reviewed (commented):
Small comments but looks good.

What do you think about grouping the output: https://docs.github.com/en/actions/learn-github-actions/workflow-commands-for-github-actions#grouping-log-lines?

millerdev reviewed (commented):
Looks good to me after Simon's comments are addressed.

> Sidenote: why is it called `python-sharded-and-javascript`? I advocate that we rename it to something like `catchall` or `everything-else`.

That test runner runs [sharded tests](https://github.com/dimagi/commcare-hq/blob/b04288b3c24077ac79bcabd8fd1240c29bb9799c/corehq/form_processor/tests/utils.py#L138-L148) as well as [JavaScript tests](https://github.com/dimagi/commcare-hq/blob/b04288b3c24077ac79bcabd8fd1240c29bb9799c/docker/run.sh#L189-L195) (in addition to a few other miscellaneous things). It's different from the other jobs because it does not run one of the `NOSE_DIVIDED_WE_RUN` groups.

joxl committed:
Improve logging clarity and add support for GitHub workflow grouping

joxl committed:
Rename make action from `migratons` to `migrations.lock`

joxl committed:
Add GitHub workflow grouping for `make requirements` test

joxl committed:
Don't write migrations log messages to STDERR

Messes up GitHub workflow output grouping

joxl committed:
Add GitHub workflow grouping for Django test suites

joxl committed:
Fix test scripts writing to STDERR

- Sending logging output to STDERR messes up GitHub workflows
- Use `curl -s` to prevent curl from writing connection metrics on STDERR

joxl committed:
Silence datadog `curl` output which was messing up GitHub workflow group

joxl committed:
Complete GitHub workflow groups for all contents of `Run tests` step

snopoke reviewed (approved):
That grouping really makes the actions log output easier to navigate!

joxl commented:
> Once you have a DB you'll first need to run `./manage.py migrate` unless you strip of the `[x]` from the output.

I just realized I never addressed this (it was relevant to my initial attempt at a new workflow action).  You bring up a good point that the current `make migrations.lock` check inherently depends on the fact that "some previous tests already ran, and therefore migrations are current".  This is safe and fine with the current workflow, and dumping the verbatim migrations list into the lock file is certainly easy. Do you think it's fine to leave the `[x]` prefix in the lockfile? I'm trying to determine if they will _ever_ serve a purpose in the lockfile, and whether or not it would be better to just remove them right now.

joxl committed:
Improve logging and CI grouping logic for container tests

millerdev reviewed (commented):
Grouping seems really nice, but I have a question. Do you know how it works when something fails inside a group? Does Actions expand that group automatically, or is it necessary to click through each group to figure out which one failed?

joxl commented:
> Do you know how it works when something fails inside a group?

I do not know, but I'll test it now.

joxl committed:
Test GitHub workflow "group failure"

joxl committed:
Test GitHub workflow "group failure"

joxl commented:
@millerdev, it does not expand any failed "groups", it only reports the non-zero return code at the end (see screenshot below).

In the case of our container tests (i.e. `scripts/docker test --noinput --stop ...`), wouldn't the failure _always_ be in the _last_ group?  The test command is just a shell command, which means GH Actions only knows pass/fail by its exit code, so GH Actions never has, and never will know exactly _where_ in the output the error occurred. If the action "script" (the target of the `run` parameter) is sane (as ours is) it will stop execution and not emit further logging output at the moment the failure happens.

![image](https://user-images.githubusercontent.com/1731363/145113111-879e7c4e-8534-4720-8b28-fbd73b228ad7.png)

joxl committed:
Revert 'Test GitHub workflow "group failure"'

This reverts commit 1485d09f28da3b23d3398450efda80f71b1b5d48.

joxl committed:
Test GitHub workflow for intermediate (i.e. not last) test group

joxl committed:
Revert "Test GitHub workflow for intermediate (i.e. not last) test group"

This reverts commit ddf795ae89c0b5ba4b6fa4f084c942ea4360f7d6.

joxl committed:
Ensure Django test suite workflow group gets closed on failure

snopoke commented:
> > Once you have a DB you'll first need to run `./manage.py migrate` unless you strip of the `[x]` from the output.
> 
> I just realized I never addressed this (it was relevant to my initial attempt at a new workflow action). You bring up a good point that the current `make migrations.lock` check inherently depends on the fact that "some previous tests already ran, and therefore migrations are current". This is safe and fine with the current workflow, and dumping the verbatim migrations list into the lock file is certainly easy. Do you think it's fine to leave the `[x]` prefix in the lockfile? I'm trying to determine if they will _ever_ serve a purpose in the lockfile, and whether or not it would be better to just remove them right now.

I think it would be better to remove the `[x] ` since that is only an indication of which migrations have been run on the environment where the command is executed and is irrelevant to what we're actually checking (are there new migrations that weren't there before). On the other hand, leaving it in forces the developer to run the migration on their local machine before updating the lock file. I don't know if there's much advantage to that.

snopoke commented:
> UPDATE: if the group fails in such a way that the ::endgroup:: message is not emitted, no group will be created and the failed "grouped section" will not be collapsible (see example job here). If the ::endgroup:: message is emitted before a group reports its failure, the group will be collapsed (see example job here).

I assume we'd always want to omit the endgroup in the case of failure right? Can the `traps` check the return code before outputting?

millerdev commented:
Approved before I saw Simon's response (that will teach me to reload). I agree, it would be nice to omit endgroup in the case of failure.

joxl commented:
> On the other hand, leaving it in forces the developer to run the migration on their local machine before updating the lock file.

They can always manually tweak the lockfile themselves if there is a real scenario where they must not run the migration locally (I don't know why this would ever be necessary, though).  I'm going to leave the `[x]`'s in the file, they can always be removed later if we decide they're problematic.

joxl committed:
Do not emit `::endgroup::` if workflow group fails

joxl committed:
Revert 'Test GitHub workflow "group failure"'

This reverts commit 19912cd1bb99d4c13ae925ba6d951cb6a73db32a.

joxl commented:
> I assume we'd always want to omit the endgroup in the case of failure right?

This is now implemented. See example (collapsed group) failure [here](https://github.com/dimagi/commcare-hq/runs/4460270860?check_suite_focus=true).
